domain/src/SharePooling/Services/DisposalProcessor.php

```diff
@@ -0,0 +1,208 @@
+        LocalDate $date,
+        QuantityBreakdown $sameDayQuantityBreakdown,
+        QuantityBreakdown $thirtyDayQuantityBreakdown,
+        Quantity $remainingQuantity,
+    ): FiatAmount {
+        assert($transactions->first() !== null);
+
+        $costBasis = $transactions->first()->costBasis->nilAmount();
+
+        return self::processSameDayAcquisitions(
+            $transactions,
+            $date,
+            $sameDayQuantityBreakdown,
+            $thirtyDayQuantityBreakdown,
+            $remainingQuantity,
+            $costBasis,
+        );
+    }
+
+    private static function processSameDayAcquisitions(
+        SharePoolingTransactions $transactions,
+        LocalDate $date,
+        QuantityBreakdown $sameDayQuantityBreakdown,
+        QuantityBreakdown $thirtyDayQuantityBreakdown,
+        Quantity $remainingQuantity,
+        FiatAmount $costBasis,
+    ): FiatAmount {
+        if ($remainingQuantity->isZero()) {
+            return $costBasis;
+        }
+
+        // Get same-day acquisitions with some quantity not matched with same-day disposals yet
+        $sameDayAcquisitions = $transactions->acquisitionsMadeOn($date)->withAvailableSameDayQuantity();
+
+        if ($sameDayAcquisitions->isEmpty()) {
+            return self::processAcquisitionsWithinThirtyDays(
+                $transactions,
+                $date,
+                $thirtyDayQuantityBreakdown,
+                $remainingQuantity,
+                $costBasis,
+            );
+        }
+
+        // Get the same-day average cost basis per unit, of all acquisitions of that day (not
+        // just the ones with available same-day quantity). We want the absolute value here,
+        // regardless of other types of matching, because same-day matching gets priority
+        $sameDayAcquisitionsAverageCostBasisPerUnit = $transactions->acquisitionsMadeOn($date)->averageCostBasisPerUnit();
+
+        assert($sameDayAcquisitionsAverageCostBasisPerUnit !== null);
+
+        // Apply this average cost basis to the disposed of asset, up to the
+        // quantity acquired that day not yet matched with same-day disposals
+        $availableSameDayQuantity = Quantity::minimum($sameDayAcquisitions->availableSameDayQuantity(), $remainingQuantity);
+        $costBasis = $costBasis->plus($sameDayAcquisitionsAverageCostBasisPerUnit->multipliedBy($availableSameDayQuantity));
+
+        // Deduct the applied quantity from the same-day acquisitions
+        $remainder = $availableSameDayQuantity;
+        foreach ($sameDayAcquisitions as $acquisition) {
+            $quantityToAssign = Quantity::minimum($remainder, $acquisition->availableSameDayQuantity());
+            $sameDayQuantityBreakdown->assignQuantity($quantityToAssign, $acquisition);
+            $acquisition->increaseSameDayQuantity($remainder);
+            $remainder =  $remainder->minus($quantityToAssign);
+            if ($remainder->isZero()) {
+                break;
+            }
+        }
```

Comment:
ℹ️ We keep track of the acquisitions' quantities matched with the disposal following the share pooling rules